### ISSUE TITLE
HTTP/2 cancellation and object null reference

### ISSUE THREAD

**Issue by JamesNK** [dotnet/corefx#38293]
HttpClient is throwing NullReferenceException when it gets RST_STREAM INTERNAL_ERROR from the server.

Wireshark log attached.

```
C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient [jamesnk/deadline-safety ≡ +0 ~3 -0 !]> dotnet run --use
_tls false --server_port 50052 --client_type httpclient --test_case timeout_on_sleeping_server
Use TLS: False
Use Test CA: False
Client type: httpclient
Server host: localhost
Server port: 50052
running timeout_on_sleeping_server
trce: Grpc.Net.Client.Internal.GrpcCall[5]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Starting deadline timeout. Duration: 00:00:00.0804225.
dbug: Grpc.Net.Client.Internal.GrpcCall[1]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Starting gRPC call. Method type: 'DuplexStreaming', URI: '/grpc.testing.TestService/FullDuplexCall'.
dbug: Grpc.Net.Client.Internal.GrpcCall[6]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Sending message.
trce: Grpc.Net.Client.Internal.GrpcCall[9]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Serialized 'Grpc.Testing.StreamingOutputCallRequest' to 27190 byte message.
trce: Grpc.Net.Client.Internal.GrpcCall[7]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Message sent.
fail: Grpc.Net.Client.Internal.GrpcCall[6]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      Error starting gRPC call.
System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Http.RedirectHandler.GetUriForRedirect(Uri requestUri, HttpResponseMessage response)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Grpc.Net.Client.Internal.GrpcCall`2.SendAsync(HttpClient client, HttpRequestMessage message) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\GrpcCall.cs:line 432

dbugUnhandled Exception: : Grpc.Net.Client.Internal.GrpcCall[8]
      => GrpcMethodType:DuplexStreaming GrpcUri:/grpc.testing.TestService/FullDuplexCall
      gRPC call canceled.
System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.) ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Http.RedirectHandler.GetUriForRedirect(Uri requestUri, HttpResponseMessage response)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Grpc.Net.Client.Internal.GrpcCall`2.SendAsync(HttpClient client, HttpRequestMessage message) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\GrpcCall.cs:line 432
   at Grpc.Net.Client.Internal.HttpContentClientStreamReader`2.MoveNextCore(CancellationToken cancellationToken) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\HttpContentClientStreamReader.cs:line 113
   at InteropTestsClient.InteropClient.RunTimeoutOnSleepingServerAsync(TestServiceClient client) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 562
   at InteropTestsClient.InteropClient.RunTestCaseAsync(IChannel channel, ClientOptions options) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 265
   at InteropTestsClient.InteropClient.Run() in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 129
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at InteropTestsClient.InteropClient.<>c.<Run>b__5_1(ClientOptions options) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 122
   at CommandLine.ParserResultExtensions.WithParsed[T](ParserResult`1 result, Action`1 action)
   at InteropTestsClient.InteropClient.Run(String[] args) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 111
   at InteropTestsClient.Program.Main(String[] args) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\Program.cs:line 25
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
   at System.Net.Http.RedirectHandler.GetUriForRedirect(Uri requestUri, HttpResponseMessage response)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Grpc.Net.Client.Internal.GrpcCall`2.SendAsync(HttpClient client, HttpRequestMessage message) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\GrpcCall.cs:line 432
   at Grpc.Net.Client.Internal.HttpContentClientStreamReader`2.MoveNextCore(CancellationToken cancellationToken) in C:\Development\Source\grpc-dotnet\src\Grpc.Net.Client\Internal\HttpContentClientStreamReader.cs:line 113
   at InteropTestsClient.InteropClient.RunTimeoutOnSleepingServerAsync(TestServiceClient client) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 562
   at InteropTestsClient.InteropClient.RunTestCaseAsync(IChannel channel, ClientOptions options) in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 265
   at InteropTestsClient.InteropClient.Run() in C:\Development\Source\grpc-dotnet\testassets\InteropTestsClient\InteropClient.cs:line 129<---
```
[httpclient-nre.zip](https://github.com/dotnet/corefx/files/3259733/httpclient-nre.zip)

**Comment by davidsh** [comment]
@wfurt

**Comment by wfurt** [comment]
can we reproduce it with interop tests @JamesNK ? 
Even if that takes a while I can trace it down.

**Comment by JamesNK** [comment]
1. git clone https://github.com/JamesNK/grpc-dotnet.git
2. git checkout jamesnk/httpclient-nre
3. grpc-dotnet\testassets\InteropTestsWebsite -> dotnet run
4. grpc-dotnet\testassets\InteropTestsClient -> dotnet run --use_tls false --serv
er_port 50052 --client_type httpclient --test_case timeout_on_sleeping_server

**Comment by wfurt** [comment]
fixed by #38226
